### PR TITLE
Add smithy4sErrorsAsScala3Unions metadata flag

### PR DESCRIPTION
**Pull request by msosnicki**

### CODE REVIEW

**Comment by msosnicki** [comment]
closes #703

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 1]
```diff
@@ -68,6 +69,8 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
       .asScala
       .toList
 
+  val rendererConfig = Renderer.Config.load(model.getMetadata().asScala.toMap) 
+
   private sealed trait DefaultRenderMode
```
Comment (msosnicki) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:74]: Should these go to the `Renderer.Config`? In current shape it would require some refactoring, as they affect `IR` only.

File: modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala [hunk 2]
```diff
@@ -22,22 +22,52 @@ import software.amazon.smithy.model.Model
...
+      expectedScalaCode: String
+  )(implicit
+      loc: Location
+  ): Unit = {
+    val scalaResults = generateScalaCode(smithySpec)
     Assertions.assertEquals(scalaResults, List(expectedScalaCode))
   }
 
+  /**
+    * Finds a code section starting with a given string and extracts a code section from it (based on indents)
```
Comment (Baccata) [inline modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala]: good stuff 👍

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 1 above)
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:74]: they should, but I'm happy for you to leave it as is for it to be tackled it in a subsequent PR (either by you or someone else).

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala [hunk 3]
```diff
@@ -560,20 +576,69 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
...
+          ) {
+            op.errors.collect { case Type.Ref(_, name) =>
+              line"case $errorName.${name}Case(e) => e"
+            }
           }
-        }
       )
   }
 
+  private def renderErrorUnion(
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/internals/Renderer.scala]: let's rename it to `renderErrorAsScala3Union` to avoid confusing potential readers, as "unions" can refer to smithy4 unions.

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala [hunk 4]
```diff
@@ -560,20 +576,69 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
...
+      shapeId: ShapeId,
+      name: NameRef,
+      alts: NonEmptyList[Alt],
+      recursive: Boolean,
+      hints: List[Hint]
+  ) = {
+    // Only Alts with UnionMember.TypeCase are valid for errors
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:614]: 👍

**Comment by Baccata** [comment]
Awesome work, besides the method rename it's a pretty good PR 👍. 

If you don't mind, I'd also like you to create an `0.17.2` section in the `CHANGELOG.md` at the root of the project and add a paragraph with this addition in it.

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala [hunk 5]
```diff
@@ -32,11 +32,24 @@ import scala.jdk.CollectionConverters._
 import Line._
 import LineSyntax.LineInterpolator
 import ToLines.lineToLines
+import java.beans.Introspector.decapitalize
```
Comment (daddykotex) [inline modules/codegen/src/smithy4s/codegen/internals/Renderer.scala]: TIL

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala (hunk 3 above)
Comment (msosnicki) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala]: Updated

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala (hunk 5 above)
Comment (msosnicki) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala]: First time in years I used `java.beans` (and hope last) :joy: Knew something has to be there is stdlib, stackoverflow to the rescue :)

Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala]: we do have an `uncapitalise` method somewhere, but this is fine, I'll probably change it later on 😄

File: CHANGELOG.md [hunk 6]
```diff
@@ -147,3 +147,11 @@ See https://github.com/disneystreaming/smithy4s/pull/649
 
 Refinements applied on list/map shapes can now produce parameterised types. This allows, for instance, to have generic
 refinements on `list` shapes that produce `cats.data.NonEmptyList` containing the same types of elements.
+
+# 0.17.2
```
Comment (Baccata) [inline CHANGELOG.md]: mmm sorry to be a pain, but would you mind putting this at the top of the file ? the latest changes in changelogs are usually at the top so that people don't have to scroll down to see what's new.

Comment (msosnicki) [reply CHANGELOG.md]: No worries, done :)

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala (hunk 5 above)
Comment (msosnicki) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala]: Got it, updated

File: modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala [hunk 7]
```diff
@@ -22,22 +22,53 @@ import software.amazon.smithy.model.Model
...
 
+  /**
+    * Finds a section starting with a given string and extracts a code section from it (based on indents)
+    * Removes empty lines too.
+    * It then asserts on expected output
+    */
+  def assertContainsSection(fileContent: String, startsWith: String)(
+      expectedSection: String
+  )(implicit loc: Location) = {
+    val lines = fileContent.lines.filter(_.trim.nonEmpty).toList.zipWithIndex
```
Comment (Baccata) [inline modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala]: CI complains about the fact that `.lines` is deprecated in 2.13 and that `.linesIterator` should be used

File: modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala [hunk 8]
```diff
@@ -22,22 +22,55 @@ import software.amazon.smithy.model.Model
...
+    val scalaResults = generateScalaCode(smithySpec)
     Assertions.assertEquals(scalaResults, List(expectedScalaCode))
   }
 
+  /**
+    * Finds a section starting with a given string and extracts a code section from it (based on indents)
+    * Removes empty lines too.
+    * It then asserts on expected output
+    */
+  def assertContainsSection(fileContent: String, startsWith: String)(
```
Comment (kubukoz) [inline modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala:50]: This function is complex enough that I'd add separate tests for it on its own ;)

File: modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala [hunk 9]
```diff
@@ -22,22 +22,55 @@ import software.amazon.smithy.model.Model
...
+      expectedSection: String
+  )(implicit loc: Location) = {
+    fileContent.linesIterator
```
Comment (kubukoz) [inline modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala]: looks like a redundant call?

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererConfigSpec.scala [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+         |}""".stripMargin
+    )
+
+    assertContainsSection(serviceCode, "sealed trait HelloError")(
+      """|sealed trait HelloError extends scala.Product with scala.Serializable {
+         |  @inline final def widen: HelloError = this
+         |}""".stripMargin
+    )
+
+    assertContainsSection(serviceCode, "object HelloError")(
```
Comment (kubukoz) [inline modules/codegen/test/src/smithy4s/codegen/internals/RendererConfigSpec.scala]: I'm just a tiny bit worried these tests may be a bit painful to maintain when an unrelated change is made...

File: modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala (hunk 9 above)
Comment (msosnicki) [reply modules/codegen/test/src/smithy4s/codegen/internals/TestUtils.scala]: Yeah, crept in somehow during my last commit. Fixed

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererConfigSpec.scala (hunk 10 above)
Comment (msosnicki) [reply modules/codegen/test/src/smithy4s/codegen/internals/RendererConfigSpec.scala]: Sure, open to alternatives how to test it, wasn't sure it's the right way anyway. 
It's basically like `fileContent.contains(expectedSection)` assert but with nicer error messages in case of failures.

Comment (Baccata) [reply modules/codegen/test/src/smithy4s/codegen/internals/RendererConfigSpec.scala]: I'm not worried about it until it proves problematic. At that point we may bring in scalameta to test, but at this point in time I'd rather not

**Comment by Baccata** [comment]
@msosnicki I'll ask you to add a documentation page as well, under the "customisation" section.

**Comment by msosnicki** [comment]
> @msosnicki I'll ask you to add a documentation page as well, under the "customisation" section.

Done, please let me know if it's ok, or I missed something

**Review by Baccata — DISMISSED** [review]
LGTM, thanks a lot for the contribution. I'll likely cut a new release later today or tomorrow.

**Comment by Baccata** [comment]
Oops had forgotten about having to manually run the CI for first-contributions.

**Comment by Baccata** [comment]
@kubukoz do you want to make another pass before I merge ?

**Comment by kubukoz** [comment]
Looks good ;)

**Review by kubukoz — APPROVED** [review]

**Review by Baccata — APPROVED** [review]
